[GStreamer] Fix readyState calculations

vivienne-w opened the pull request:
#### 738525821b033b3a945e32e3a9c4255af0af595a
<pre>
[GStreamer] Fix readyState calculations
<a href="https://bugs.webkit.org/show_bug.cgi?id=248217">https://bugs.webkit.org/show_bug.cgi?id=248217</a>

Reviewed by Philippe Normand.

On some platforms, the audio sink is acting as a fake sink while the decoder fetches data
from the pipeline and transfers it to the SoC drivers, even in READY and PAUSED states,
meaning we cannot rely on buffering messages anymore.

However, on that platform the audio sinks implements buffering queries, that are currently
issued to the entire pipeline, which on some platforms may yield misleading results
if some random element implements buffering queries and receives that query first.

Thus this patch first queries the audio sink, before trying the video sink and finally
the pipeline itself.

Original author: Pawel Lampe &lt;pawel.lampe@gmail.com&gt;
See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/975">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/975</a>

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateStates): Send buffering query to audio/video sinks first

Canonical link: <a href="https://commits.webkit.org/257066@main">https://commits.webkit.org/257066@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9132c98bca80f0a70e20d8bed8399822ed5911a2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97770 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7009 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30947 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107261 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167531 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101717 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7438 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35784 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90152 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103907 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103409 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84403 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32542 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89207 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/998 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/990 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22138 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4853 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5816 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2245 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41541 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2337,9 +2337,10 @@ void MediaPlayerPrivateGStreamer::updateStates()
                 GRefPtr<GstQuery> query = adoptGRef(gst_query_new_buffering(GST_FORMAT_PERCENT));
 
                 m_isBuffering = m_bufferingPercentage < 100;
-                if (gst_element_query(m_pipeline.get(), query.get())) {
```
Shouldn't we fallback to this if the query wasn't handled by sinks?

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2337,9 +2337,10 @@ void MediaPlayerPrivateGStreamer::updateStates()
 
                 m_isBuffering = m_bufferingPercentage < 100;
-                if (gst_element_query(m_pipeline.get(), query.get())) {
+                if ((m_audioSink && gst_element_query(m_audioSink.get(), query.get())) || (m_videoSink && gst_element_query(m_videoSink.get(), query.get())) || gst_element_query(m_pipeline.get(), query.get())) {
```
Just a nit, this is a quite a long line, maybe it would be better to separate it in three for readability. Please remember that in this coding style the `||` would go at the beginning of the next line.

With those changes, LGTM, but I'll let @philn have the final word as he did the initial review.

philn replied:
+1 to break this long line :)

vivienne-w replied:
like this?
```c
if ((m_audioSink && gst_element_query(m_audioSink.get(), query.get()))
    || (m_videoSink && gst_element_query(m_videoSink.get(), query.get()))
    || gst_element_query(m_pipeline.get(), query.get()))
{
    gboolean isBuffering = m_isBuffering;
    gst_query_parse_buffering_percent(query.get(), &isBuffering, nullptr);
    GST_TRACE_OBJECT(pipeline(), "[Buffering] m_isBuffering forcefully updated from %d to %d", m_isBuffering, isBuffering);
    m_isBuffering = isBuffering;
}
```

philn replied:
Yep!

eocanha commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2337,9 +2337,13 @@ void MediaPlayerPrivateGStreamer::updateStates()
+                if ((m_audioSink && gst_element_query(m_audioSink.get(), query.get()))
+                    || (m_videoSink && gst_element_query(m_videoSink.get(), query.get()))
+                    || gst_element_query(m_pipeline.get(), query.get()))
+                {
```
Taking advantage of having to resubmit the patch because the "Reviewed by" line was absent, please fix the breakage of the [braces-blocks rule of the coding style](https://webkit.org/code-style-guidelines/#braces-blocks) and place the opening brace at the end of the previous line, instead of on its own line. Braces on their own lines are [reserved for method/function definitions](https://webkit.org/code-style-guidelines/#braces-function).

Thanks in advance!

calvaris replied:
What is weird is that the style bot didn't catch this...

vivienne-w committed:
[GStreamer] Fix readyState calculations
https://bugs.webkit.org/show_bug.cgi?id=248217

Reviewed by Philippe Normand.

On some platforms, the audio sink is acting as a fake sink while the decoder fetches data
from the pipeline and transfers it to the SoC drivers, even in READY and PAUSED states,
meaning we cannot rely on buffering messages anymore.

However, on that platform the audio sinks implements buffering queries, that are currently
issued to the entire pipeline, which on some platforms may yield misleading results
if some random element implements buffering queries and receives that query first.

Thus this patch first queries the audio sink, before trying the video sink and finally
the pipeline itself.

Original author: Pawel Lampe <pawel.lampe@gmail.com>
See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/975

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateStates): Send buffering query to audio/video sinks first

Canonical link: https://commits.webkit.org/257066@main
